File: web/js/renderers/earthRenderer.js
import { drawSphere, drawLabel } from './sphereRenderer.js';

// Renders Earth with day/night terminator, city lights, and observer dot.

export function drawEarth(ctx, cx, cy, radius, lightDirX, lightDirY, earthRotationDeg = 0) {
  // Base blue sphere (day side)
  drawSphere(ctx, cx, cy, radius, [40, 80, 200], lightDirX, lightDirY);

  // Green continent hints
  ctx.fillStyle = 'rgba(50,160,60,0.39)';
  ctx.beginPath();
  ctx.ellipse(
    cx - radius * 0.1, cy - radius * 0.25,
    radius * 0.2, radius * 0.15, 0, 0, Math.PI * 2
  );
  ctx.fill();

  ctx.beginPath();
  ctx.ellipse(
    cx + radius * 0.05, cy + radius * 0.2,
    radius * 0.2, radius * 0.15, 0, 0, Math.PI * 2
  );
  ctx.fill();

  // Day/Night terminator shadow
  drawDayNightTerminator(ctx, cx, cy, radius, lightDirX, lightDirY);

  // City lights on the dark side
  drawCityLights(ctx, cx, cy, radius, lightDirX, lightDirY);

  // Atmosphere glow
  const atmoGrad = ctx.createRadialGradient(cx, cy, 0, cx, cy, radius * 1.15);
  atmoGrad.addColorStop(0.75, 'rgba(100,180,255,0)');
  atmoGrad.addColorStop(0.9, 'rgba(100,180,255,0.12)');
  atmoGrad.addColorStop(1, 'rgba(100,180,255,0)');
  ctx.fillStyle = atmoGrad;
  ctx.beginPath();
  ctx.arc(cx, cy, radius * 1.15, 0, Math.PI * 2);
  ctx.fill();

  // Observer dot on the equator
  drawObserver(ctx, cx, cy, radius, earthRotationDeg, lightDirX, lightDirY);
}

function drawDayNightTerminator(ctx, cx, cy, radius, lightDirX, lightDirY) {
  ctx.save();

  // Clip to Earth circle
  ctx.beginPath();
  ctx.arc(cx, cy, radius, 0, Math.PI * 2);
  ctx.clip();

  const angle = Math.atan2(-lightDirY, -lightDirX);
  const perpAngle1 = angle + Math.PI / 2;

  // Shadow path — dark half facing away from light
  ctx.fillStyle = 'rgba(5,5,30,0.71)';
  ctx.beginPath();
  ctx.moveTo(
    cx + Math.cos(perpAngle1) * radius,
    cy + Math.sin(perpAngle1) * radius
  );
  ctx.arc(
    cx, cy, radius,
    perpAngle1,
    perpAngle1 - Math.PI,
    true
  );
  ctx.lineTo(
    cx + Math.cos(perpAngle1) * radius,
    cy + Math.sin(perpAngle1) * radius
  );
  ctx.closePath();
  ctx.fill();

  // Soft terminator glow (twilight zone)
  const perpAngle2 = angle - Math.PI / 2;
  ctx.strokeStyle = 'rgba(255,150,80,0.12)';
  ctx.lineWidth = radius * 0.15;
  ctx.beginPath();
  ctx.moveTo(
    cx + Math.cos(perpAngle1) * radius,
    cy + Math.sin(perpAngle1) * radius
  );
  ctx.lineTo(
    cx + Math.cos(perpAngle2) * radius,
    cy + Math.sin(perpAngle2) * radius
  );
  ctx.stroke();

  ctx.restore();
}

function drawCityLights(ctx, cx, cy, radius, lightDirX, lightDirY) {
  ctx.save();
  ctx.beginPath();
  ctx.arc(cx, cy, radius * 0.95, 0, Math.PI * 2);
  ctx.clip();

  const cities = [
    [-0.15, -0.1], [0.2, -0.2], [0.1, 0.15],
    [-0.25, 0.2], [0.3, -0.05], [-0.1, -0.3],
    [0.0, 0.1], [-0.3, -0.05], [0.15, 0.25],
  ];

  for (const [dx, dy] of cities) {
    const ccx = cx + dx * radius;
    const ccy = cy + dy * radius;
    const toCityX = dx;
    const toCityY = dy;
    const dot = toCityX * lightDirX + toCityY * lightDirY;

    if (dot < -0.1) {
      const brightness = Math.min(1, Math.abs(dot));
      ctx.fillStyle = `rgba(255,220,100,${brightness * 120 / 255})`;
      ctx.beginPath();
      ctx.arc(ccx, ccy, 1.2, 0, Math.PI * 2);
      ctx.fill();
    }
  }

  ctx.restore();
}

function drawObserver(ctx, cx, cy, radius, earthRotationDeg, lightDirX, lightDirY) {
  const rotRad = (earthRotationDeg * Math.PI) / 180.0;

  const obsX = cx - radius * 0.9 * Math.cos(rotRad);
  const obsY = cy + radius * 0.3 * Math.sin(rotRad);

  const sinComponent = Math.sin(rotRad);
  if (sinComponent > -0.3) {
    const relX = (obsX - cx) / radius;
    const dot = relX * lightDirX;
    const isDay = dot > 0;

    ctx.fillStyle = isDay
      ? 'rgba(255,255,0,0.86)'
      : 'rgba(255,150,50,0.86)';
    ctx.beginPath();
    ctx.arc(obsX, obsY, 3.5, 0, Math.PI * 2);
    ctx.fill();

    drawLabel(
      ctx,
      isDay ? 'You' : 'zzZ',
      obsX, obsY - 10,
      10,
      isDay ? [255, 255, 200] : [200, 200, 255]
    );
  }
}

File: MoonSimulation/Renderers/EarthRenderer.cs
using SkiaSharp;

namespace MoonSimulation.Renderers;

/// <summary>
/// Renders Earth with day/night terminator, city lights, and observer dot.
/// </summary>
public static class EarthRenderer
{
    /// <summary>
    /// Draws Earth with a blue/green sphere, day/night terminator, city lights, and observer dot.
    /// </summary>
    public static void DrawEarth(SKCanvas canvas, SKPoint center, float radius, SKPoint lightDir,
        double earthRotationDeg = 0)
    {
        // Base blue sphere (day side)
        SphereRenderer.DrawSphere(canvas, center, radius, new SKColor(40, 80, 200), lightDir, 0.2f);

        // Green continent hints
        using var continentPaint = new SKPaint
        {
            IsAntialias = true,
            Color = new SKColor(50, 160, 60, 100),
            Style = SKPaintStyle.Fill
        };
        canvas.DrawOval(
            new SKRect(center.X - radius * 0.3f, center.Y - radius * 0.4f,
                       center.X + radius * 0.1f, center.Y - radius * 0.1f),
            continentPaint);
        canvas.DrawOval(
            new SKRect(center.X - radius * 0.15f, center.Y + radius * 0.05f,
                       center.X + radius * 0.25f, center.Y + radius * 0.35f),
            continentPaint);

        // Day/Night terminator shadow
        DrawDayNightTerminator(canvas, center, radius, lightDir);

        // City lights on the dark side
        DrawCityLights(canvas, center, radius, lightDir);

        // Atmosphere glow
        using var atmoPaint = new SKPaint { IsAntialias = true };
        using var atmoShader = SKShader.CreateRadialGradient(
            center, radius * 1.15f,
            new SKColor[] { SKColors.Transparent, new SKColor(100, 180, 255, 30), SKColors.Transparent },
            new float[] { 0.75f, 0.9f, 1f },
            SKShaderTileMode.Clamp);
        atmoPaint.Shader = atmoShader;
        canvas.DrawCircle(center, radius * 1.15f, atmoPaint);

        // Observer dot on the equator
        DrawObserver(canvas, center, radius, earthRotationDeg, lightDir);
    }

    /// <summary>
    /// Draws the dark (night) half of Earth facing away from the Sun.
    /// </summary>
    public static void DrawDayNightTerminator(SKCanvas canvas, SKPoint center, float radius,
        SKPoint lightDir)
    {
        canvas.Save();
        using var clipPath = new SKPath();
        clipPath.AddCircle(center.X, center.Y, radius);
        canvas.ClipPath(clipPath);

        float angle = MathF.Atan2(-lightDir.Y, -lightDir.X);

        using var shadowPaint = new SKPaint
        {
            IsAntialias = true,
            Color = new SKColor(5, 5, 30, 180),
            Style = SKPaintStyle.Fill
        };

        using var shadowPath = new SKPath();
        float perpAngle1 = angle + MathF.PI / 2;
        float perpAngle2 = angle - MathF.PI / 2;

        shadowPath.MoveTo(
            center.X + MathF.Cos(perpAngle1) * radius,
            center.Y + MathF.Sin(perpAngle1) * radius);

        shadowPath.ArcTo(
            new SKRect(center.X - radius, center.Y - radius,
                       center.X + radius, center.Y + radius),
            (perpAngle1 * 180f / MathF.PI),
            -180, false);

        shadowPath.LineTo(
            center.X + MathF.Cos(perpAngle1) * radius,
            center.Y + MathF.Sin(perpAngle1) * radius);
        shadowPath.Close();

        canvas.DrawPath(shadowPath, shadowPaint);

        // Soft terminator glow (twilight zone)
        using var twilightPaint = new SKPaint
        {
            IsAntialias = true,
            Style = SKPaintStyle.Stroke,
            StrokeWidth = radius * 0.15f,
            Color = new SKColor(255, 150, 80, 30)
        };
        canvas.DrawLine(
            center.X + MathF.Cos(perpAngle1) * radius,
            center.Y + MathF.Sin(perpAngle1) * radius,
            center.X + MathF.Cos(perpAngle2) * radius,
            center.Y + MathF.Sin(perpAngle2) * radius,
            twilightPaint);

        canvas.Restore();
    }

    /// <summary>
    /// Draws tiny city light dots on the night side of Earth.
    /// </summary>
    public static void DrawCityLights(SKCanvas canvas, SKPoint center, float radius,
        SKPoint lightDir)
    {
        canvas.Save();
        using var clipPath = new SKPath();
        clipPath.AddCircle(center.X, center.Y, radius * 0.95f);
        canvas.ClipPath(clipPath);

        using var lightPaint = new SKPaint
        {
            IsAntialias = true,
            Color = new SKColor(255, 220, 100, 140)
        };

        var cities = new (float dx, float dy)[]
        {
            (-0.15f, -0.1f), (0.2f, -0.2f), (0.1f, 0.15f),
            (-0.25f, 0.2f), (0.3f, -0.05f), (-0.1f, -0.3f),
            (0.0f, 0.1f), (-0.3f, -0.05f), (0.15f, 0.25f),
        };

        foreach (var (dx, dy) in cities)
        {
            float cx = center.X + dx * radius;
            float cy = center.Y + dy * radius;

            float toCityX = (cx - center.X) / radius;
            float toCityY = (cy - center.Y) / radius;
            float dot = toCityX * lightDir.X + toCityY * lightDir.Y;

            if (dot < -0.1f) // On the dark side
            {
                float brightness = Math.Min(1f, Math.Abs(dot));
                lightPaint.Color = new SKColor(255, 220, 100, (byte)(brightness * 120));
                canvas.DrawCircle(cx, cy, 1.2f, lightPaint);
            }
        }

        canvas.Restore();
    }

    /// <summary>
    /// Draws a small observer indicator on Earth's equator.
    /// </summary>
    public static void DrawObserver(SKCanvas canvas, SKPoint center, float radius,
        double earthRotationDeg, SKPoint lightDir)
    {
        double rotRad = earthRotationDeg * Math.PI / 180.0;

        float obsX = center.X + radius * 0.9f * (float)Math.Cos(rotRad + Math.PI);
        float obsY = center.Y + radius * 0.3f * (float)Math.Sin(rotRad + Math.PI);

        float sinComponent = (float)Math.Sin(rotRad + Math.PI);
        if (sinComponent > -0.3f) // Visible hemisphere (with tilt)
        {
            float relX = (obsX - center.X) / radius;
            float dot = relX * lightDir.X;
            bool isDay = dot > 0;

            using var obsPaint = new SKPaint
            {
                IsAntialias = true,
                Color = isDay ? new SKColor(255, 255, 0, 220) : new SKColor(255, 150, 50, 220)
            };
            canvas.DrawCircle(obsX, obsY, 3.5f, obsPaint);

            SphereRenderer.DrawLabel(canvas, isDay ? "You" : "zzZ",
                new SKPoint(obsX, obsY - 10), 10,
                isDay ? new SKColor(255, 255, 200) : new SKColor(200, 200, 255));
        }
    }
}
